### PR TITLE
Refactor resolve_dependencies

### PR DESCRIPTION
**Pull request by matklad**
This moves `resolve_dependencies` function from `compile.rs` into the `resolve.rs`, where it more reasonably belongs.

### CODE REVIEW

File: src/cargo/ops/resolve.rs [hunk 1]
```diff
@@ -1,30 +1,101 @@
...
+
+fn resolve_with_registry(ws: &Workspace, registry: &mut PackageRegistry, write_lockfile: bool)
+                         -> CargoResult<Resolve> {
     let prev = ops::load_pkg_lockfile(ws)?;
     let resolve = resolve_with_previous(registry, ws,
                                         Method::Everything,
                                         prev.as_ref(), None, &[])?;
 
-    // Avoid writing a lockfile if we are `cargo install`ing a non local package.
-    if ws.current_opt().map(|pkg| pkg.package_id().source_id().is_path()).unwrap_or(true) {
```
Comment (matklad) [inline src/cargo/ops/resolve.rs]: This is the only actual change in the code that is in this PR. Instead of this scary condition, we basically using `source.is_some()` [other there](https://github.com/rust-lang/cargo/pull/3413/files#diff-e3d98804dece93349c90df7540234488R41). I believe these are equivalent.

File: src/cargo/ops/resolve.rs [hunk 2]
```diff
@@ -1,30 +1,101 @@
...
+    }).map(|s| s.to_string()).collect::<Vec<String>>();
+
+    let mut registry = PackageRegistry::new(ws.config())?;
+
+    let mut write_lockfile = true;
+    if let Some(source) = source {
+        if let Some(root_package) = ws.current_opt() {
+            // Avoid writing a lockfile if we are `cargo install`ing a non local package.
+            write_lockfile = false;
+            registry.add_preloaded(root_package.package_id().source_id(), source);
```
Comment (alexcrichton) [inline src/cargo/ops/resolve.rs]: Looking back at this, I think we may want to refactor this a little bit along with `write_lockfile` above. I'm slightly worried about two things here:

1. There may not be a correlation between `Box<Source>` and the current package's source id.
2. Passing in `source` is a historical optimization to not re-read the current package again, so while currently it implies `cargo install` that may not always be true.

Perhaps we can fix (1) by adding a `Source::source_id` method? (should be trivial to add). And for (2) can we check to see of (2) is not a path dependency, and then that implies `write_lockfile = false`?

**Comment by alexcrichton** [comment]
Looks great, thanks!

File: src/cargo/ops/resolve.rs (hunk 2 above)
Comment (matklad) [inline src/cargo/ops/resolve.rs]: I'll need some time to wrap my head around these `Source`s, `Registry`s and `RegistrySource`s :) 

Can we perhaps remember this info during workspace creation, instead of threading it as parameters via several function? I am thinking about something along the lines of

```
struct Workspace {
  kind: WorkspaceKind
}

enum WorkspaceKind { 
  Local,
  Temporary(Box<Source>)
}
```

Comment (matklad) [inline src/cargo/ops/resolve.rs]: Ah, my idea won't work nicely, because `add_preloaded` consumes the `Source`, and this is reasonable.

Comment (alexcrichton) [inline src/cargo/ops/resolve.rs]: Yeah I wouldn't mind threading this info somehow through the `Workspace`. I guess my point though was that whether or not we have a `Box<Source>` is orthogonal to whether we're doing a `cargo install`, they just happen to coincide today.

But to clarify:

* `Registry` - a trait for objects where packages can be *queried*. Given a dependency description, a registry can return a list of package candidates that it contains.
* `Source` - a trait which extends `Registry` above, a source represents a concrete way of loading packages from some source. This encapsulates information such as updating the source (e.g. updating the index or fetching a git repo) and downloading crates (e.g. a literal download or just a loading of a crate). 
* `RegistrySource` - this is actually just a concrete implementation of `Source` for the crates.io registry.

I... agree that our naming is quite confusing though! I had to reword this a few times in writing it...

**Comment by matklad** [comment]
Ok, so Sources now expose their ids. I think there may be some potential edge cases around `ReplacedSource` and precise  vs imprecise sources, but the tests pass!

Not writing the lockfile if the package is not path dependency seems a little fishy, because lockfile belongs to the workspace and not to the package. Remembering if the workspace is temporary seems clearer.

**Comment by alexcrichton** [comment]
@bors: r+

Looks great to me, thanks!

**Comment by matklad** [comment]
Found one more instance where it is possible to get rid of `SourceId` now.
